contracts/root/TokenPredicates/BurnableERC20Predicate.sol

```diff
@@ -0,0 +1,115 @@
+    using RLPReader for RLPReader.RLPItem;
+
+    bytes32 public constant MANAGER_ROLE = keccak256("MANAGER_ROLE");
+    bytes32 public constant TOKEN_TYPE = keccak256("BurnableERC20");
+    // Standard ERC20 transfer event signature : keccak256("Transfer(address,address,uint256)")
+    bytes32 public constant TRANSFER_EVENT_SIG = 0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef;
+    // When you've burned tokens on L2 & want to do same on L1 : keccak256("Burn(address,uint256)"), signature
+    // event needs to be emitted
+    bytes32 public constant BURN_EVENT_SIG = 0xcc16f5dbb4873280815c1ee09dbd06736cffcc184412cf7a71a0fdb75d397ca5;
+
+    // Offchain L1 clients can keep track of how many tokens getting deposited, by whom etc.
+    event LockedBurnableERC20(
+        address indexed depositor,
+        address indexed depositReceiver,
+        address indexed rootToken,
+        uint256 amount
+    );
+
+    constructor() public {}
+
+    function initialize(address _owner) external initializer {
+        _setupContractId("BurnableERC20Predicate");
+        _setupRole(DEFAULT_ADMIN_ROLE, _owner);
+        _setupRole(MANAGER_ROLE, _owner);
+    }
+
+    /**
+     * @notice Lock burnable ERC20 tokens for deposit, callable only by manager, when depositing from L1 to L2
+     * @param depositor Address who wants to deposit tokens
+     * @param depositReceiver Address (address) who wants to receive tokens on child chain
+     * @param rootToken Token which gets deposited
+     * @param depositData ABI encoded amount
+     */
+    function lockTokens(
+        address depositor,
+        address depositReceiver,
+        address rootToken,
+        bytes calldata depositData
+    ) external override only(MANAGER_ROLE) {
+        uint256 amount = abi.decode(depositData, (uint256));
+
+        emit LockedBurnableERC20(depositor, depositReceiver, rootToken, amount);
+        IBurnableERC20(rootToken).transferFrom(
+            depositor,
+            address(this),
+            amount
+        );
+    }
+
+    /**
+     * @notice Validates log signature, from and to address
+     * then burns whole withdrawn amount
+     *
+     * @param rootToken L1 token which gets withdrawn
+     * @param log Valid ERC20 burn log from child chain
+     */
+    function exitTokens(
+        address,
+        address rootToken,
+        bytes memory log
+    ) public override only(MANAGER_ROLE) {
+        RLPReader.RLPItem[] memory logRLPList = log.toRlpItem().toList();
+        RLPReader.RLPItem[] memory logTopicRLPList = logRLPList[1].toList(); // topics
+
+        // User who's withdrawing
+        address withdrawer = address(logTopicRLPList[1].toUint());
+
+        // Checking if user wanted to get fund back on L1, after burning it on L2
+        if(bytes32(logTopicRLPList[0].toUint()) == TRANSFER_EVENT_SIG) {
+            
+            require(address(logTopicRLPList[2].toUint()) == address(0), "BurnableERC20Predicate: INVALID_RECEIVER");
```

Comment:
logTopicRLPList[2] is the `to` field,  right,? I am probably missing something but how do you prevent anyone from using a normal transfer on L2 here ?

Reply:
Respective event signature is different.

```js
 if(bytes32(logTopicRLPList[0].toUint()) == TRANSFER_EVENT_SIG) {}
```

Reply:
Sorry I do not follow, the TRANSFER_EVENT_SIG is that of ERC20, right ?
if so `logTopicRLPList[2] ` is `to` and so it means any such event can be used to bring a token back to L1 ?

Reply:
Yes, that's the purpose of `emit Transfer(0x[a-zA-Z0-9]{40}, 0x0{40}, [0-9]{1,})` on L2.

Reply:
ok, so that's an issue. you do not want anyone to take the token back on L1 if it was transfered to anyone on L2

Reply:
Not really, if it's explicitly burnt using `emit Transfer(0x[a-zA-Z0-9]{40}, 0x0{40}, [0-9]{1,})`, it's considered to be eligible for exit from L2.

One can transfer to anyone on L2, supply doesn't change on L2, total supply is assumed to be same, as total sum of tokens of L2 are locked on L1, but anyone can exit with those, given they have ownership of it on L2 & burn it with aforementioned signature.

**REAL burning** is different, which is what this PR adds.

Reply:
I misread the code actually.
But my main issue with the burn event remain: it complicate the code and require you to add further event.
If you prefers to continue this route, I would propose a simpler PR with the ChainExit event which have all the functionality of this one with only one single extra event.  and as mentioned, the extra functionality is not only real burning,, but a generic mechanism for exit that allow to carry data across.